sky/skylet/providers/gcp/node_provider.py

```diff
@@ -387,6 +388,10 @@ def get_command_runner(
             "use_internal_ip": use_internal_ip,
         }
         if docker_config and docker_config["container_name"] != "":
+            if "docker_login_config" in self.provider_config:
+                docker_config["docker_login_config"] = DockerLoginConfig.from_dict(
+                    self.provider_config["docker_login_config"]
+                )
```

Comment:
Was wondering if we can directly have the docker login config set in the docker section in the ray yaml, instead of having it in the provider config?

Reply:
ray has a schema validator for their YAML file and it doesn't allow a customized field in `docker`. Here is one traceback I received:
```bash
Traceback (most recent call last):
  File "/tmp/skypilot_ray_up_92hwrqy0.py", line 76, in <module>
    sdk.create_or_update_cluster('/home/memory/.sky/generated/cbl-test-docker-login.yml', **{'no_restart': True})
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/ray/autoscaler/sdk/sdk.py", line 48, in create_or_update_cluster
    use_login_shells=True,
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/ray/autoscaler/_private/commands.py", line 279, in create_or_update_cluster
    config = _bootstrap_config(config, no_config_cache=no_config_cache)
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/ray/autoscaler/_private/commands.py", line 373, in _bootstrap_config
    validate_config(config)
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/ray/autoscaler/_private/util.py", line 162, in validate_config
    raise e from None
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/ray/autoscaler/_private/util.py", line 160, in validate_config
    jsonschema.validate(config, schema)
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/jsonschema/validators.py", line 1121, in validate
    raise error
jsonschema.exceptions.ValidationError: Additional properties are not allowed ('docker_login_config' was unexpected)

Failed validating 'additionalProperties' in schema['properties']['docker']:
    {'additionalProperties': False,
     'description': 'Docker configuration. If this is specified, all setup '
                    'and start commands will be executed in the container.',
     'properties': {'container_name': {'default': 'ray_docker',
                                       'type': 'string'},
                    'disable_automatic_runtime_detection': {'default': False,
                                                            'description': 'disable '
                                                                           'Ray '
                                                                           'from '
                                                                           'automatically '
                                                                           'using '
                                                                           'the '
                                                                           'NVIDIA '
                                                                           'runtime '
                                                                           'if '
                                                                           'available',
                                                            'type': 'boolean'},
                    'disable_shm_size_detection': {'default': False,
                                                   'description': 'disable '
                                                                  'Ray '
                                                                  'from '
                                                                  'automatically '
                                                                  'detecting '
                                                                  '/dev/shm '
                                                                  'size '
                                                                  'for the '
                                                                  'container',
                                                   'type': 'boolean'},
                    'head_image': {'description': 'image for head node, '
                                                  'takes precedence over '
                                                  "'image' if specified",
                                   'type': 'string'},
                    'head_run_options': {'description': 'head specific run '
                                                        'options, appended '
                                                        'to run_options',
                                         'type': 'array'},
                    'image': {'default': 'rayproject/ray:latest',
                              'description': 'the docker image name',
                              'type': 'string'},
                    'pull_before_run': {'description': 'run `docker pull` '
                                                       'first',
                                        'type': 'boolean'},
                    'run_options': {'description': 'shared options for '
                                                   'starting head/worker '
                                                   'docker',
                                    'type': 'array'},
                    'use_podman': {'default': False,
                                   'description': "Use 'podman' command in "
                                                  "place of 'docker'",
                                   'type': 'boolean'},
                    'worker_image': {'description': 'analogous to '
                                                    'head_image',
                                     'type': 'string'},
                    'worker_run_options': {'description': 'analogous to '
                                                          'head_run_options',
                                           'type': 'array'}},
     'type': 'object'}
```

Reply:
Ahh, got it. That is annonying. Seems they disabled `additionalProperties` for the docker config but not for the provider's. In that case, might be fine to keep it in provider, but can we have a comment mentioning this? https://github.com/ray-project/ray/blob/d2fc4823126927b2c54f89ec72fa3d24b442e6a3/python/ray/autoscaler/ray-schema.json#L227